File: parser_combinator_rs/src/combinator.rs
use crate::foundation::{Ctx, Result};
use std::cell::Cell;

pub type ParserFn<T> = Box<dyn Fn(&Ctx) -> Result<T>>;

pub fn any<T>(parsers: Vec<ParserFn<T>>) -> impl Fn(&Ctx) -> Result<T> {
  fn _parser<T>(ctx: &Ctx, parsers: &Vec<ParserFn<T>>) -> Result<T> {
    let mut err = ctx.failure("any".to_string());

    for parser in parsers {
      let res = parser(ctx);
      let failure = match res {
        Ok(success) => return Ok(success),
        Err(failure) => failure,
      };
      if failure.index() > err.index() {
        err = failure;
      }
    }
    Err(err)
  }

  move |ctx| _parser(ctx, &parsers)
}

// pub fn sequence<T: Clone>(parsers: Vec<ParserFn<T>>) -> impl Fn(&Ctx) -> Result<Vec<T>> {
//   fn _parser<T: Clone>(ctx: &Ctx, parsers: &Vec<ParserFn<T>>) -> Result<Vec<T>> {
//     let mut values: Vec<T> = vec![];
//     let mut next_ctx = ctx.to_owned();
//     for parser in parsers {
//       let success = parser(&next_ctx)?;
//       next_ctx = success.ctx().to_owned();
//       let val = success.val();
//       values.push(val);
//     }
//     Ok(next_ctx.success(values))
//   }
//   move |ctx| _parser(ctx, &parsers)
// }

pub fn many<T: Clone>(parser: impl Fn(&Ctx) -> Result<T>) -> impl Fn(&Ctx) -> Result<Vec<T>> {
  move |ctx| {
    let mut values: Vec<T> = vec![];
    let mut next_ctx = ctx.to_owned();
    loop {
      match parser(&next_ctx) {
        Err(_) => break,
        Ok(success) => {
          next_ctx = success.ctx().to_owned();
          let val = success.val();
          values.push(val);
        }
      }
    }
    Ok(next_ctx.success(values))
  }
}

pub fn delimited<T: Clone, L, R>(
  left: impl Fn(&Ctx) -> Result<L>,
  parser: impl Fn(&Ctx) -> Result<T>,
  right: impl Fn(&Ctx) -> Result<R>,
) -> impl Fn(&Ctx) -> Result<T> {
  move |ctx| {
    let l_res = left(ctx)?;
    let mut next_ctx = l_res.ctx();
    let res = parser(next_ctx)?;
    next_ctx = res.ctx();
    let r_res = right(next_ctx)?;
    next_ctx = r_res.ctx();
    Ok(next_ctx.success(res.val()))
  }
}

pub fn delimited_left<T: Clone, L>(
  left: impl Fn(&Ctx) -> Result<L>,
  parser: impl Fn(&Ctx) -> Result<T>,
) -> impl Fn(&Ctx) -> Result<T> {
  move |ctx| {
    let l_res = left(ctx)?;
    let next_ctx = l_res.ctx();
    parser(next_ctx)
  }
}

pub fn separated<T: Clone>(
  separator: impl Fn(&Ctx) -> Result<String>,
  parser: impl Fn(&Ctx) -> Result<T>,
) -> impl Fn(&Ctx) -> Result<Vec<T>> {
  let is_firt = Cell::new(true);
  let skip_first = move |ctx: &Ctx| {
    if is_firt.get() {
      is_firt.set(false);
      Ok(ctx.success("".to_owned()))
    } else {
      separator(ctx)
    }
  };

  many(delimited_left(skip_first, parser))
}

//fn map<T: Clone, R>(
//  parser: impl Fn(&Ctx) -> Result<T>,
//  op: impl Fn(T) -> R,
//) -> impl Fn(&Ctx) -> Result<R> {
//  move |ctx| {
//    let result = parser(ctx);
//    match result {
//      Ok(success) => {
//        let ctx = success.ctx();
//        let val = op(success.val());
//        Ok(ctx.success(val))
//      }
//      Err(failure) => Err(failure),
//    }
//  }
//}

//#[cfg(test)]
//mod test {
//  use crate::foundation::Ctx;
//}

File: parser_combinator_ts/src/combinator.ts
import { Failure, Parser } from "./foundation";

export function any<T>(parsers: Parser<T>[]): Parser<T> {
  return (ctx) => {
    if (parsers.length === 0) return ctx.failure("any");
    let failure: Failure | null = null;
    for (const parser of parsers) {
      const res = parser(ctx);
      if (res.success) return res;
      if (!failure || failure.ctx.index < res.ctx.index) {
        failure = res as Failure;
      }
    }
    return failure;
  };
}

export function many<T>(parser: Parser<T>): Parser<T[]> {
  return (ctx) => {
    const values: T[] = [];
    let nextCtx = ctx;
    while (true) {
      const res = parser(nextCtx);
      if (!res.success) {
        if (nextCtx.index == res.ctx.index) break;
        return res as Failure;
      }
      values.push(res.value);
      nextCtx = res.ctx;
    }
    return nextCtx.success(values);
  };
}

export function sequence<T>(parsers: Parser<T>[]): Parser<T[]> {
  return (ctx) => {
    let values: T[] = [];
    let nextCtx = ctx;
    for (const parser of parsers) {
      const res = parser(nextCtx);
      if (!res.success) return res as Failure;
      values.push(res.value);
      nextCtx = res.ctx;
    }
    return nextCtx.success(values);
  };
}

export function map<A, B>(parser: Parser<A>, fn: (val: A) => B): Parser<B> {
  return (ctx) => {
    const res = parser(ctx);
    return !res.success ? (res as Failure) : res.ctx.success(fn(res.value));
  };
}

export function delimited<T, L, R>(
  left: Parser<L>,
  parser: Parser<T>,
  right: Parser<R>
): Parser<T> {
  return (ctx) => {
    const leftRes = left(ctx);
    if (!leftRes.success) return leftRes as Failure;
    const res = parser(leftRes.ctx);
    if (!res.success) return res as Failure;
    const rightRes = right(res.ctx);
    if (!rightRes.success) return rightRes as Failure;
    return rightRes.ctx.success(res.value);
  };
}

export function delimitedLeft<T>(
  left: Parser<string>,
  parser: Parser<T>
): Parser<T> {
  return (ctx) => {
    const leftRes = left(ctx);
    if (!leftRes.success) return leftRes as Failure;
    return parser(leftRes.ctx);
  };
}
